torch/_inductor/cudagraph_trees.py

```diff
@@ -0,0 +1,863 @@
+        assert len(tree_manager.roots) > 0, "expected at least one use"
+        root = next(tree_manager.get_roots())
+        self.graph = root.graph
+        for t in live_tensors:
+            self.live_tensors_count += 1
+            weakref.finalize(t, self.finalize_tensor)
+
+    def get_tree_manager(self) -> CUDAGraphTreeManager:
+        with self.lock:
+            if self.tree_manager is None:
+                self.tree_manager = CUDAGraphTreeManager()
+            return self.tree_manager
+
+
+local = threading.local()
+local.tree_manager_container = TreeManagerContainer()
+
+# We need to register this as an object that will be copied over as TLS when new
+# threads are created in autograd
+torch._C._stash_obj_in_tls("tree_manager_container", local.tree_manager_container)
+
+
+def get_container():
+    if hasattr(local, "tree_manager_container"):
+        return local.tree_manager_container
+    assert torch._C._is_key_in_tls("tree_manager_container")
+    return torch._C._get_obj_in_tls("tree_manager_container")
+
+
+def cudagraphify_impl(model, inputs, static_input_idxs=()):
+    manager = get_container().get_tree_manager()
+    return manager.add_function(model, inputs, static_input_idxs)
+
+
+def is_live(weak_ref):
+    if weak_ref is None:
+        return False
+    return weak_ref() is not None
+
+
+class TensorWeakRef(object):
+    """
+    Wrapper around a weak ref of a Tensor that handles the _fix_weakref() call required
+    when unwrapping a Tensor weakref.
+    """
+
+    ref: WeakRef[Tensor]
+
+    def __init__(self, tensor: Tensor):
+        assert isinstance(tensor, Tensor)
+        self.ref = weakref.ref(tensor)
+
+    def __call__(self):
+        out = self.ref()
+        if out is None:
+            return out
+        assert isinstance(out, Tensor)
+        out._fix_weakref()
+        return out
+
+
+def is_cuda_tensor(x):
+    return isinstance(x, torch.Tensor) and x.device.type == "cuda"
+
+
+class CUDAGraphNode(object):
+    """
+    A single recording of a function into a CUDA Graph. Recordings of CUDA Graphs share a single memory pool
+    and are structured into a tree, where there is a single recording that can precede it (parent) and multiple
+    subsequent recordings that may follow (children). A node will have no parent if it is the first recording
+    in a tree; i.e., when it is first recorded, there are no live tensors from a previous recording which
+    would force a dependency.
+
+    On first recording, all of the live tensors in the current CUDA Graph Node path will be
+    reflected in the corresponding private pool. On subsequent executions, the caching allocator
+    is unaffected when the graph is replayed.
+
+    In order to support recording a subsequent cuda graph recording after execution of this graph,
+    we checkpoint the state of the memory pool so that it may later be resumed.
+
+    See [setCheckpointPoolState] for further explanation.
```

Comment:
Link to the diagram too here?

Reply:
By diagram, do you mean the ascii diagram or the diagram that we made on excaldiraw ?

Reply:
excalidraw